File: organized-skills/1password-skill/tools-python/src/op_env/op_env_create.py
"""op-env-create — Create environment item in 1Password (Python SDK version).

Usage:
    op-env-create <name> <vault> [--from-file <.env>] [--tags <tags>] [KEY=VALUE ...]
"""

from __future__ import annotations

import argparse
import asyncio
import os
import sys

from onepassword.types import (
    ItemCategory,
    ItemCreateParams,
    ItemField,
    ItemFieldType,
    ItemSection,
)

from .utils import (
    get_client,
    item_exists,
    log,
    parse_env_file,
    parse_inline_vars,
    resolve_vault_id,
)

HELP_EPILOG = """\
examples:
  # Create with inline variables
  uv run op-env-create my-app-dev Development API_KEY=xxx DB_HOST=localhost

  # Create from .env file
  uv run op-env-create my-app-prod Production --from-file .env.prod

  # Combine file and inline (inline overrides file)
  uv run op-env-create azure-config Shared --from-file .env EXTRA_KEY=value

  # With custom tags
  uv run op-env-create secrets DevOps --tags "env,production,api" KEY=value
"""


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="op-env-create",
        description="Create environment item in 1Password",
        epilog=HELP_EPILOG,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("positionals", nargs="*", help="<name> <vault> [KEY=VALUE ...]")
    parser.add_argument("--from-file", dest="from_file", help="Import variables from .env file")
    parser.add_argument(
        "--tags", default="environment", help="Comma-separated tags (default: environment)"
    )
    return parser


async def main() -> None:
    parser = build_parser()
    args = parser.parse_args()

    # Separate name, vault, and KEY=VALUE from positionals
    name = ""
    vault = ""
    env_args: list[str] = []

    for arg in args.positionals:
        if "=" in arg:
            env_args.append(arg)
        elif not name:
            name = arg
        elif not vault:
            vault = arg

    if not name or not vault:
        log.error("Missing required arguments: name and vault")
        parser.print_help()
        sys.exit(1)

    client = await get_client()
    vault_id = await resolve_vault_id(client, vault)

    # Check if item already exists
    if await item_exists(client, vault_id, name):
        log.error(f"Item '{name}' already exists in vault '{vault}'")
        log.info("Use op-env-update to modify existing environment")
        sys.exit(1)

    # Collect variables
    variables: dict[str, str] = {}

    # Load from file if specified
    if args.from_file:
        if not os.path.isfile(args.from_file):
            log.error(f"File not found: {args.from_file}")
            sys.exit(1)
        log.info(f"Loading variables from: {args.from_file}")
        with open(args.from_file) as f:
            variables.update(parse_env_file(f.read()))

    # Add/override with inline variables
    variables.update(parse_inline_vars(env_args))

    if not variables:
        log.error("No environment variables provided")
        log.info("Use --from-file <.env> or provide KEY=value pairs")
        sys.exit(1)

    log.info(f"Creating environment: {name} in vault: {vault}")
    log.info(f"Variables ({len(variables)}): {' '.join(variables.keys())}")

    # Build the item using the SDK
    # Create fields in a "variables" section with CONCEALED type
    section = ItemSection(id="variables", title="variables")
    fields = []
    for key, value in variables.items():
        fields.append(
            ItemField(
                id=key,
                title=key,
                value=value,
                field_type=ItemFieldType.CONCEALED,
                section_id="variables",
            )
        )

    params = ItemCreateParams(
        title=name,
        category=ItemCategory.APICREDENTIALS,
        vault_id=vault_id,
        tags=args.tags.split(","),
        sections=[section],
        fields=fields,
    )

    try:
        await client.items.create(params)
        log.info(f"Environment '{name}' created successfully")
        print("\nVariables stored:")
        for key in variables:
            print(f"  - {key}")
        print("\nUsage:")
        print("  # Export to .env file")
        print(f"  uv run op-env-export '{name}' '{vault}' > .env")
        print()
        first_key = next(iter(variables))
        print("  # Read single variable")
        print(f"  op read 'op://{vault}/{name}/variables/{first_key}'")
    except Exception as exc:
        log.error(f"Failed to create environment: {exc}")
        sys.exit(1)


def main_sync() -> None:
    asyncio.run(main())


if __name__ == "__main__":
    main_sync()

File: organized-skills/1password-skill/tools/src/op-env-create.ts
#!/usr/bin/env bun
/**
 * op-env-create - Create environment item in 1Password
 *
 * Usage:
 *   op-env-create <name> <vault> [--from-file <.env>] [key=value...]
 *
 * Examples:
 *   op-env-create my-app-dev Development API_KEY=xxx DB_HOST=localhost
 *   op-env-create my-app-prod Production --from-file .env.prod
 */

import { parseArgs } from "util";
import {
  log,
  ensureAuth,
  itemExists,
  parseEnvFile,
  parseInlineVars,
  createEnvironment,
} from "./utils";

const HELP = `
op-env-create - Create environment item in 1Password

USAGE:
    op-env-create <name> <vault> [OPTIONS] [key=value...]

ARGUMENTS:
    <name>      Name/title for the environment item
    <vault>     Vault to store the item in

OPTIONS:
    --from-file <path>   Import variables from .env file
    --tags <tags>        Comma-separated tags (default: environment)
    --help               Show this help message

EXAMPLES:
    # Create with inline variables
    op-env-create my-app-dev Development API_KEY=xxx DB_HOST=localhost

    # Create from .env file
    op-env-create my-app-prod Production --from-file .env.prod

    # Combine file and inline (inline overrides file)
    op-env-create azure-config Shared --from-file .env EXTRA_KEY=value

    # With custom tags
    op-env-create secrets DevOps --tags "env,production,api" KEY=value
`;

async function main() {
  const { values, positionals } = parseArgs({
    args: Bun.argv.slice(2),
    options: {
      "from-file": { type: "string" },
      tags: { type: "string", default: "environment" },
      help: { type: "boolean", short: "h" },
    },
    allowPositionals: true,
  });

  if (values.help) {
    console.log(HELP);
    process.exit(0);
  }

  // Parse positionals: name, vault, and key=value pairs
  const envVars: string[] = [];
  let name = "";
  let vault = "";

  for (const arg of positionals) {
    if (arg.includes("=")) {
      envVars.push(arg);
    } else if (!name) {
      name = arg;
    } else if (!vault) {
      vault = arg;
    }
  }

  if (!name || !vault) {
    log.error("Missing required arguments: name and vault");
    console.log(HELP);
    process.exit(1);
  }

  // Ensure authenticated
  await ensureAuth();

  // Check if item already exists
  if (await itemExists(name, vault)) {
    log.error(`Item '${name}' already exists in vault '${vault}'`);
    log.info("Use op-env-update to modify existing environment");
    process.exit(1);
  }

  // Collect variables
  const vars = new Map<string, string>();

  // Load from file if specified
  if (values["from-file"]) {
    const filePath = values["from-file"];
    const file = Bun.file(filePath);

    if (!(await file.exists())) {
      log.error(`File not found: ${filePath}`);
      process.exit(1);
    }

    log.info(`Loading variables from: ${filePath}`);
    const content = await file.text();
    const fileVars = parseEnvFile(content);
    for (const [k, v] of fileVars) {
      vars.set(k, v);
    }
  }

  // Add/override with inline variables
  const inlineVars = parseInlineVars(envVars);
  for (const [k, v] of inlineVars) {
    vars.set(k, v);
  }

  if (vars.size === 0) {
    log.error("No environment variables provided");
    log.info("Use --from-file <.env> or provide KEY=value pairs");
    process.exit(1);
  }

  log.info(`Creating environment: ${name} in vault: ${vault}`);
  log.info(`Variables (${vars.size}): ${Array.from(vars.keys()).join(" ")}`);

  // Create the environment
  const success = await createEnvironment(name, vault, vars, values.tags);

  if (success) {
    log.info(`Environment '${name}' created successfully`);
    console.log("\nVariables stored:");
    for (const key of vars.keys()) {
      console.log(`  - ${key}`);
    }
    console.log("\nUsage:");
    console.log(`  # Export to .env file`);
    console.log(`  op-env-export '${name}' '${vault}' > .env`);
    console.log();
    const firstKey = vars.keys().next().value;
    console.log(`  # Read single variable`);
    console.log(`  op read 'op://${vault}/${name}/variables/${firstKey}'`);
  } else {
    log.error("Failed to create environment");
    process.exit(1);
  }
}

main().catch((err) => {
  log.error(err.message);
  process.exit(1);
});
